core: add cephx keyType api

BlaineEXE opened the pull request:
Add the CephX `keyType` API that allows users to specify what type of CephX key Rook should create or rotate to.

<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

BlaineEXE committed:
core: add cephx keyType api

Add the CephX `keyType` API that allows users to specify what type of
CephX key Rook should create or rotate to.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE commented:
Note: the CSV generated for this branch/PR is from the master branch that has some modifications to what is currently in the `release-*` branches (from the work done to allow the script to run on macos). If `keyType` is backported, each release branch will need to drop this CSV update and redo the csv generation

BlaineEXE committed:
csv: update csv for keyType API updates

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE commented:
TODO:
- [x] include https://github.com/rook/rook/pull/17052
- [x] finish key rotation for PVC-based OSDs

BlaineEXE committed:
build: use ceph dev image for keyType - TEMPORARY

TEMPORARY! DO NOT INCLUDE IN BACKPORTS!

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE committed:
ci: test keyType with ceph images

Modify CI tests to use Ceph images containing keyType capabilities.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE commented on tests/framework/installer/ceph_installer.go:
```diff
@@ -47,7 +47,7 @@ const (
 	// test with the current development versions
 	squidDevelTestImage    = "quay.ceph.io/ceph-ci/ceph:squid"
-	tentacleDevelTestImage = "quay.ceph.io/ceph-ci/ceph:tentacle"
+	tentacleDevelTestImage = "quay.ceph.io/ceph-ci/ceph:wip-pdonnell-testing-20260415.142642-centos-stream9" // actually main
```
TESTING ONLY!

BlaineEXE commented on tests/framework/installer/ceph_installer.go:
```diff
@@ -99,7 +99,7 @@ func ReturnCephVersion() cephv1.CephVersionSpec {
 	default:
 		// Default to the latest stable version
-		return SquidVersion
+		return TentacleDevelVersion
```
TESTING ONLY!

BlaineEXE commented on pkg/operator/ceph/controller/controller_utils.go:
```diff
@@ -194,8 +194,9 @@ func canIgnoreHealthErrStatusInReconcile(cephCluster cephv1.CephCluster, control
 	}
 
 	allowedErrStatus := map[string]struct{}{
-		"MDS_ALL_DOWN":     {},
-		"MGR_MODULE_ERROR": {},
+		"MDS_ALL_DOWN":                   {},
+		"MGR_MODULE_ERROR":               {},
+		"AUTH_INSECURE_SERVICE_KEY_TYPE": {}, // rook can reconcile cephx keys to clear this
 	}
```
Small but important part of these changes. I think this is good to keep, but noting for discussion.

parth-gr commented on pkg/daemon/ceph/client/auth.go:
```diff
@@ -117,9 +126,12 @@ func AuthGetCaps(context *clusterd.Context, clusterInfo *ClusterInfo, name strin
 	logger.Infof("rotating ceph auth key %q", name)
 	args := []string{"auth", "rotate", name}
+	if keyType != "" {
+		args = append(args, KeyTypeFlag, keyType)
```
Does this needed for external cluster script rotation?

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1070,6 +1081,9 @@ def create_cephCSIKeyring_user(self, user, generation):
             "format": "json",
         }
 
+        if self._arg_parser.cephx_key_type != "":
```
The CephX keys are created by the Python script itself, so we should do the creation/rotation based on the key type https://github.com/rook/rook/blob/master/deploy/examples/create-external-cluster-resources.py#L1066

BlaineEXE replied:
We had an old conversation many months back, and we don't plan to do that (right now). The new flag is documented that it only takes affect when new keys are being created or when rotation is enabled.

This will be a little odd for users, but it makes the scripting significantly easier, and much easier to backport. This should be a relatively rare need, so I don't think it needs to be perfect.

We can also work to improve upon this later if we get feedback about issues.

BlaineEXE committed:
core: fix LeastUptodateDaemonVersion corner cases

Fix an error return bug and missing corner case logic in
LeastUptodateDaemonVersion.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE committed:
core: add workaround for OSD upgrade plus cephx rotation

Add a workaround for a Ceph bug when upgrading from a
non-AES256K-supporting version to an AES256K-supporting version when
keys are rotated in the same reconcile.

Ceph does intend to fix this bug, but it will take an unknown amount
of time to release. Rook can maintain this workaround in the meantime to
prevent users who perform both at once from losing OSD-to-OSD
communication during the upgrade.

The workaround is implemented using a per-CephCluster global enable flag
internally. This is non-ideal but allows Rook to implement the temporary
workaround without more dramatic alterations to the many reconciles in
which cephx key rotation is managed.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE committed:
core: fix mon cephx status log content

Fix the "updating mon daemon cephx status" log to properly reflect the
mon status. Mgr status was inadvertently used prior.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -59,6 +59,10 @@ const (
 
+// admin key rotation ends with a rook restart, so return a reconcile error after to help
+// prevent rook from continuing reconcile when restart is imminent
+var errOsdUpgradeCephxRotationWorkaroundRestart = fmt.Errorf("cephx key rotation re-enablement requires the current cluster reconcile to restart")
```
This is the same pattern used for OSDs, so I borrowed it to keep consistency

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -947,3 +957,65 @@ func (c *cluster) deleteBootstrapKeys() {
+		// Now that the least-versioned OSD supports it, reload the manager to force all reconcilers
+		// to reassess whether keys should/can be rotated.
+		logger.Infof("restarting rook operator after cephx key rotation is re-enabled for cluster in namespace %q", c.Namespace)
+		reloadManagerFunc()
```
Same. I think this is a similar case where we really do (outside of multi-CephCluster envs) want nearly all resources to be re-reconciled, and this is a quick way to do that. The less-ideal-ness in those environments should be a minor inconvenience 🤞

BlaineEXE commented on pkg/operator/ceph/config/keyring/cephx.go:
```diff
@@ -39,17 +39,49 @@ var CephAuthRotateSupportedVersion = version.CephVersion{Major: 19, Minor: 2, Ex
 //nolint:gosec // G101: this is not hardcoded credentials
 const CephxKeyIdentifierAnnotation = "cephx-key-identifier"
 
+var cephxKeyRotationAllowedForCluster = map[string]bool{}
```
Great catch! sync.Map seems unnecessarily complicated for this, so I just added a mutex var.

BlaineEXE commented on Documentation/Storage-Configuration/Advanced/cephx-key-rotation.md:
```diff
@@ -138,6 +177,91 @@ status:
+key type. This process is as simple as specifying `keyType: aes` for the required component. This
+will rotate the key again, and the new key will be of type `aes`.
+
+## Known issues
```
Overall, I agree. In this case, the per-NFS-export limitation is one of the few deep issues that some users will certainly run into during their rotations, so I thought I might try to save some time responding to repeat issue submissions about it

BlaineEXE committed:
manifest: update toolbox to watch keyring updates

Update the toolbox script and toolbox manifest to watch for updates to
the cluster keyring.

Signed-off-by: Blaine Gardner <blaine.gardner@ibm.com>
